Dashboard/utilities/email_verification.py

```diff
@@ -0,0 +1,64 @@
+'''utility functions to add unverified email to ses and send verification emails.'''
+
+from os import environ as CONFIG
+import logging
+
+from dotenv import load_dotenv
+
+from boto3 import client as boto_client
+from botocore.client import BaseClient
+from botocore.exceptions import ClientError
+import mypy_boto3_ses.client as ses_client
+
+def get_ses_client(config: dict) -> ses_client:
+    '''Returns an ses client from a configuration.'''
+    return boto_client(
+        'ses',
+        aws_access_key_id = config["ACCESS_KEY"],
+        aws_secret_access_key = config['SECRET_ACCESS_KEY'],
+        region_name='eu-west-2'
+    )
+
+def is_ses(boto_ses_client: ses_client) -> bool:
+    '''Returns true if ses client else false.'''
+    return (isinstance(boto_ses_client, BaseClient)
+            and boto_ses_client._service_model.service_name == 'ses') #pylint: disable=protected-access
+
+def send_verification_email(boto_ses_client: ses_client, email: str) -> dict[bool, str]:
+    '''adds email as unverified email to ses and send a verification email.
+    Returns a dictionary containing fields:
+            success    : True if sending email verification was a success, else False
+            reason     : If success is False, return a reason for failure
+            error      : If success is False, return an error object (only if types are correct)
+            request_id : If success is True, return a request id for that email verification.
+    '''
+    if not isinstance(email, str):
+        logging.error('send_verification_email passed `email` argument not of type str.')
+        return {'success': False, 'reason': 'bad email type, email must be of type str.'}
+    if not is_ses(boto_ses_client):
+        logging.error('send_verification_email client is not a boto3 ses client.')
+        return {'success': False, 'reason': 'client is not a boto3 ses client.'}
+    logging.info('Sending email verification...')
+    try:
+        response = boto_ses_client.verify_email_identity(EmailAddress=email)
+    except ClientError as e:
+        if e.response.get('Error').get('Code') == 'InvalidClientTokenId':
```

Comment:
Does using 2 .get() methods consecutively work if the first one fails?

Reply:
It does not. That is a good point, I will look for this explicitly

Reply:
I've just fixed it - it was a really good spot